testctrl/docs/specs/container_image_names.md

```diff
@@ -0,0 +1,115 @@
+
+3.  The `USER` specifies the username of the person who built the image.
+
+4.  The `TIMESTAMP` is the time when the image was built, represented in
+    seconds since the unix epoch.
+
+5.  The `COMPONENT_KIND` specifies what the image contains. For example, this
+    may be `"driver"` for the test orchestrator and `"worker"` for a client or
+    server.
+
+6.  The `LANGUAGE_CODE` indicates the programming language of the component.
+    This differentiates two components that may be different but built
+    simultaneously. It is omitted for component types which only have 1
+    language, like the `"driver"`.
+
+7.  The only relevant version/tag is `"latest"`, because only one image lives in
+    a repository.
+
+### Benefits
+
+The registry prefix (1) permits adherence to the naming conventions of
+registries other than DockerHub. For example, Google Container Registry (GCR)
+uses the prefix `gcr.io/<PROJECT_NAME>` on all repositories.
+
+The word "benchmarks" helps organize images. Projects may have many images, and
+this keyword shows which are related to performance tests. In GCR, these
+repositories are displayed in a nested form, like a subdirectory.
+
+The user and component kind allow developers to quickly see who built the image
+and what the image contains. This also provides a way for automated scripts to
+set different TTLs based on the component kind or user. For example, the images
+of a continuous integration bot may have different lifetimes than those of a
+developer.
+
+By including the timestamp in the repository name, an automated tool does not
+need to make a request to view the tags for each repository. It may be able to
+make a single request before deletion: [`GET /v2/_catalog`][catalog].
+
+[catalog]: https://docs.docker.com/registry/spec/api/#catalog
+
+### Drawbacks
+
+Repository names can grow fairly long, and Docker used to cap names at 30
+characters. However, this cap has been increased to 255. See [moby/moby#10392].
+
+[moby/moby#10392]: https://github.com/moby/moby/issues/10392
+
+To keep names minimal, the number of seconds since the unix epoch is used as the
+timestamp. This will remain just 10 characters until the year 2286. Usernames
+are generally capped at a few dozen characters, and languages use their shortest
+recognizable form:
+
+| Official Name        | Language Code |
+| :------------------: | :-----------: |
+| C++                  | cc            |
+| C# (C core)          | cs            |
+| C# (.NET)            | dotnet        |
+| Dart                 | dart          |
+| Go                   | go            |
+| Java                 | java          |
+| JavaScript (web)     | js            |
+| JavaScript (Node.js) | node          |
+| Kotlin               | kt            |
+| Objective-C          | objc          |
+| PHP                  | php           |
+| Python               | py            |
+| Ruby                 | rb            |
+| Swift                | swift         |
```

Comment:
I went ahead and standardized codes here based, since they may be added in the future. For instance, server-side kotlin/swift/dart, etc. This way the spec won't have to be updated. Most are based on accepted file extension, except for node.js and objective-c which would have been `js` and `m`. :octocat: